### DISCUSSION TITLE
Generating the concept menue of the tree editor depending on the state of the current emf model

### DISCUSSION THREAD

**Discussion by Deel96** [eclipse-emfcloud/emfcloud#173] [Coffee Editor] [answered]
Hi,

I was wondering if it's possible to create the context menue options on runtime.
If I'm e.g. in the coffee-editor example I have to define on compile time a childrenMapping/TreeEditor.ChildrenDescriptor in coffee-model.ts as well as the general structure in coffee-schema.

In theory I could use the modelserver api to fetch the needed values, but since the values are exported via code I can't run my fetch request before the module is getting loaded. There is a sync-fetch package on npm which I could use in theory but when I tested it the modelserver wasn't ready and gave me an empty model.

constructed example:
If I have  different types of components defined in my emf model, I only want to show them in the context menue if there is a entry on the property "activatedComponents" on the machine.
Example:

ecore model:
- Machine
- ControlUnit
- DipTray
- BrewingUnit

machine.xmi / machine.coffee:
 - eClass: ...#Machine
 - activatedComponents: ['...#DipTray','...#ControlUnit']

Contextmenue on Machine:
- Control Unit
- DipTry
- BrewingUnit
To keep it simple the contextmenue entries of the machine can be hardcoded.
  
Contextmenue on Control Unit:
- DipTray
- ControlUnit

Idealley these entries were already created dynamicly on startup

--- **Adding BrewingUnit to the list activatedComponents** ---

machine.xmi / machine.coffee:
- eClass: ...#Machine
- activatedComponents: ['...#DipTray','...#ControlUnit', 'BrewingUnit']
  
Contextmenue on Control Unit:
- DipTray
- ControlUnit
- BrewingUnit

I could life with defining a complete schema (with BrewingUnit included) at compile time and only dynamicly create the context menue entries.
Is something like that possible at the moment?

Thank you and Best Regards

Edit: Better formulation

**Comment by ndoschek** [comment, marked as answer]
Hi @Deel96,

To create children for a selected node via the `+` button (I am not aware of any context menu so far),
the tree editor fetches the information from the [`coffee-model-service#getChildrenMapping`](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-model-service.ts#L118).
This is currently hard-coded, the children mapping is defined in the [`coffee-model.ts`](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-model.ts#L53).

I suppose you could override the `getChildrenMapping` method in the coffee-model-service by calculating your custom mapping.
Perhaps you will also have to update the method in [`coffee-node-factory#hasCreatableChildren`](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-node-factory.ts#L89) as this also uses the static childrenMapping.

HTH, BR Nina

**Reply by Deel96** [reply]
HI @ndoschek,
with context menue I meant the `+` button :).

In the mean time I tried basicly what you suggested.
I changed  the childrenMapping inside [coffee-model.ts](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-model.ts#L53) to a function which makes a syncronous get request to the model sever (ugly but for testing okay). 

<details>

```ts
export function getChildrenMapping(): Map<string, TreeEditor.ChildrenDescriptor[]> {

const init = (): any => {
            const data = syncFetch('http://localhost:8081/api/v1/models', {
            }).json();
            return (Object.values(data?.data)[0] as any)?.components || undefined;
        };

        const components= init();
        
        const components = [
         components[0].eClass,
         "hardcodedURI"
        ];
        
         const childrenMapping: Map<string, TreeEditor.ChildrenDescriptor[]> = new Map([
            [
                Type.Machine, [
                    {
                        property: 'children',
                        children: components
                    }
                ]
            ],
...

return childrenMapping;
}
```

</details>

Fetching it at this place synchronously doesn't always work, depending on if the modelserver is already running or not (race condition).
Depending on if the modelserver is available I only get the hardcoded option.
But even when the model server is running I have to refresh the webpage  because otherwise the options won't be created (see below: util.ts in theia-tree-browser)

I changed [coffee-model-service#getChildrenMapping](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-model-service.ts#L118) accordingly.

I also updated [coffee-node-factory#hasCreatableChildren](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-node-factory.ts#L89) to use the function.

I think the problem is that [coffee-node-factory#hasCreatableChildren](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/web/coffee-editor-extension/src/browser/coffee-tree/coffee-node-factory.ts#L89) which is executed when the `+` button is clicked returns a boolean and doesn't update the mapping / consumes a new mapping which might be created.

At startup the childrenmapping seems to be used here for creation of the `+` menue: [util.ts](https://github.com/eclipse-emfcloud/theia-tree-editor/blob/31ad3317decd7e74be6b4f7304ee12a7fa37d465/theia-tree-editor/src/browser/util.ts#L84) but it doesn't get run again and I couldn't find a way to trigger it. (I thought about redux but redux seems not be exposed in theia)

Do you have an idea how I could get the updated context menue without refreshing and maybe even without a sync api request?

Thank you and Best Regards

**Reply by ndoschek** [reply]
Hi @Deel96,

As far as I can see the creation and registration of the add commands for the `+` menu happens directly on startup in the [`BaseTreeEditorContribution`](https://github.com/eclipse-emfcloud/theia-tree-editor/blob/31ad3317decd7e74be6b4f7304ee12a7fa37d465/theia-tree-editor/src/browser/tree-editor-contribution.ts).

This is due to the `MenuContribution` concept of Theia: see [`menu-model-registry`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/menu/menu-model-registry.ts).
On startup (`onStart()`) the commands for the menu are registered. However, this registry allows to (un-)register commands also outside that hook.

You could try to implement a listener in  the `CoffeeTreeEditorContribution` that updates the commands on model updates via the model server client.
For example, the `CoffeeTreeEditorWidget` updates the tree via `onIncrementalUpdateListener`. You could re-use this concept and update the add commands on that event as well. I'm not sure if you can only update the commands or have to un- and re-register them (and how expensive that is), but basically I think for a test showcase it should work.

Hope that helps and works out, best regards,
Nina
